src/NzbDrone.Mono/Disk/DiskProvider.cs

```diff
@@ -189,6 +194,46 @@ public override bool TryCreateHardLink(string source, string destination)
             }
         }
 
+        public override bool TryCreateRefLink(string source, string destination)
+        {
+            var source_fd = Syscall.open(source, OpenFlags.O_RDONLY);
+            if (source_fd == -1) {
+                Logger.Trace("RefLink: Cannot open source file '{0}'", source);
+                return false;
+            }
+
+            var dest_fd = Syscall.open(destination, OpenFlags.O_WRONLY | OpenFlags.O_CREAT, FilePermissions.S_IRUSR | FilePermissions.S_IWUSR);
+            if (dest_fd == -1) {
+                Syscall.close(source_fd);
+                Logger.Trace("RefLink: Cannot open dest file '{0}'", destination);
+                return false;
+            }
+
+            try
+            {
+                var result = FiCloneIoCtl(dest_fd, FICLONE, source_fd);
+                if (result == 0) {
+                    Syscall.close(source_fd);
+                    Syscall.close(dest_fd);
+                    return true;
+                }
+                var error = Stdlib.GetLastError();
+                Logger.Debug("Cannot issue FICLONE ('{0}' -> '{1}') due to '{2}'", Stdlib.strerror(error));
+            }
+            catch (EntryPointNotFoundException e)
+            {
+                Logger.Debug(e, "Platform does not support FICLONE");
+            }
+            catch (DllNotFoundException e)
+            {
+                Logger.Debug(e, "No libc.so on this platform; macOS currently unsupported for reflink");
```

Comment:
We would probably want to check for platform compatibility at the beginning of this function rather than catching the error on incompatible platforms. Don't know if we'd want to explicitly check for windows or not if that's handled by stuff in `NzbDrone.Windows`.

Reply:
I supposed that the Windows "check" was just the `NzbDrone.Windows` class/implementation.

I'm afraid that checking platform compatibility is very specific, and this catch is probably the cleanest way... There are several conditions to check if we really want this:

- Per Mono platform: Linux vs macOS vs FreeBSD vs Solaris
- Since a specific (Linux) kernel version
- Since a specific filesystem version
- Different platforms have different APIs/syscalls

So quite frankly, that'd add a lot of code and a lot of test cases :-)

Reply:
It looks like `cp` takes the [same approach](https://unix.stackexchange.com/questions/318705/cp-with-reflink-flag-how-to-determine-if-reflink-is-possible): try a reflink and see if it fails.  So I guess fair enough to do the same here.

However at this point I think we should be able to call `OsInfo.IsLinux` and avoid the try/catch on macOS.

Reply:
I think you gave to catch that `DllNotFoundException` anyhow, so I don't think it really matters that much.